flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,454 @@
+			ResultPartitionTest.createPartitionDeploymentDescriptor(ResultPartitionType.BLOCKING);
+		final ExecutionAttemptID eid1 = taskResultPartitionDescriptor.getShuffleDescriptor().getResultPartitionID().getProducerId();
+
+		final TaskDeploymentDescriptor taskDeploymentDescriptor =
+			TaskExecutorSubmissionTest.createTaskDeploymentDescriptor(
+				jobId,
+				"job",
+				eid1,
+				new SerializedValue<>(new ExecutionConfig()),
+				"Sender",
+				1,
+				0,
+				1,
+				0,
+				new Configuration(),
+				new Configuration(),
+				TestInvokable.class.getName(),
+				Collections.singletonList(taskResultPartitionDescriptor),
+				Collections.emptyList(),
+				Collections.emptyList(),
+				Collections.emptyList(),
+				0);
+
+		final TaskSlotTable taskSlotTable = createTaskSlotTable();
+
+		final TaskExecutorLocalStateStoresManager localStateStoresManager = new TaskExecutorLocalStateStoresManager(
+			false,
+			new File[]{tmp.newFolder()},
+			Executors.directExecutor());
+
+		final JobAwareShuffleEnvironmentImpl<?> jobAwareShuffleEnvironment = new JobAwareShuffleEnvironmentImpl<>(new NettyShuffleEnvironmentBuilder().build());
+
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder()
+			.setTaskSlotTable(taskSlotTable)
+			.setTaskStateManager(localStateStoresManager)
+			.setShuffleEnvironment(jobAwareShuffleEnvironment)
+			.build();
+
+		final CompletableFuture<Void> taskFinishedFuture = new CompletableFuture<>();
+
+		final TestingJobMasterGateway jobMasterGateway = new TestingJobMasterGatewayBuilder()
+			.setRegisterTaskManagerFunction((s, location) -> CompletableFuture.completedFuture(new JMTMRegistrationSuccess(ResourceID.generate())))
+			.setOfferSlotsFunction((resourceID, slotOffers) -> CompletableFuture.completedFuture(slotOffers))
+			.setUpdateTaskExecutionStateFunction(taskExecutionState -> {
+				if (taskExecutionState.getExecutionState() == ExecutionState.FINISHED) {
+					taskFinishedFuture.complete(null);
+				}
+				return CompletableFuture.completedFuture(Acknowledge.get());
+			})
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices);
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			final String jobMasterAddress = "jm";
+			rpc.registerGateway(jobMasterAddress, jobMasterGateway);
+
+			// inform the task manager about the job leader
+			taskManagerServices.getJobLeaderService().addJob(jobId, jobMasterAddress);
+			jobManagerLeaderRetriever.notifyListener(jobMasterAddress, UUID.randomUUID());
+
+			final Deadline deadline = Deadline.fromNow(Duration.ofSeconds(5));
+			while (!taskManagerServices.getJobManagerTable().contains(jobId) && deadline.hasTimeLeft()) {
+				Thread.sleep(50);
+			}
+			assertTrue(taskManagerServices.getJobManagerTable().contains(jobId));
+
+			taskSlotTable.allocateSlot(0, jobId, taskDeploymentDescriptor.getAllocationId(), Time.seconds(60));
```

Comment:
Also problematic wrt concurrency

Reply:
Moreover, this makes assumption about how the `TaskExecutor` internally works. I think such a test setup is brittle and should be avoided.